src/assets/TokensController.ts

```diff
@@ -213,102 +260,237 @@ export class TokensController extends BaseController<
+        const previousIndex = newTokens.indexOf(previousEntry);
+        newTokens[previousIndex] = newEntry;
       } else {
-        tokens.push(newEntry);
+        newTokens.push(newEntry);
       }
 
       const newIgnoredTokens = ignoredTokens.filter(
         (tokenAddress) => tokenAddress.toLowerCase() !== address.toLowerCase(),
       );
-      const { newAllTokens, newAllIgnoredTokens } = this._getNewAllTokensState(
-        tokens,
-        newIgnoredTokens,
+      const newDetectedTokens = detectedTokens.filter(
+        (token) => token.address.toLowerCase() !== address.toLowerCase(),
       );
+      const { newAllTokens, newAllIgnoredTokens, newAllDetectedTokens } =
+        this._getNewAllTokensState({
+          newTokens,
+          newIgnoredTokens,
+          newDetectedTokens,
+        });
 
       this.update({
+        tokens: newTokens,
+        ignoredTokens: newIgnoredTokens,
+        detectedTokens: newDetectedTokens,
         allTokens: newAllTokens,
-        tokens,
         allIgnoredTokens: newAllIgnoredTokens,
-        ignoredTokens: newIgnoredTokens,
+        allDetectedTokens: newAllDetectedTokens,
       });
-      return tokens;
+      return newTokens;
     } finally {
       releaseLock();
     }
   }
 
   /**
-   * Adds a batch of tokens to the stored token list.
+   * Add a batch of tokens.
    *
-   * @param tokensToAdd - Array of Tokens to be added or updated.
-   * @returns Current token list.
+   * @param tokensToImport - Array of tokens to import.
    */
-  async addTokens(tokensToAdd: Token[]): Promise<Token[]> {
+  async addTokens(tokensToImport: Token[]) {
     const releaseLock = await this.mutex.acquire();
-    const { tokens, ignoredTokens } = this.state;
+    const { tokens, detectedTokens, ignoredTokens } = this.state;
+    const importedTokensMap: { [key: string]: true } = {};
+    // Used later to dedupe imported tokens
+    const newTokensMap = tokens.reduce((output, current) => {
+      output[current.address] = current;
+      return output;
+    }, {} as { [address: string]: Token });
 
     try {
-      tokensToAdd = await Promise.all(
-        tokensToAdd.map(async (token) => {
-          token.isERC721 = await this._detectIsERC721(token.address);
-          return token;
-        }),
+      tokensToImport.forEach((tokenToAdd) => {
+        const { address, symbol, decimals, image, aggregators } = tokenToAdd;
+        const checksumAddress = toChecksumHexAddress(address);
+        const formattedToken: Token = {
+          address: checksumAddress,
+          symbol,
+          decimals,
+          image,
+          aggregators,
+        };
+        newTokensMap[address] = formattedToken;
+        importedTokensMap[address.toLowerCase()] = true;
+        return formattedToken;
+      });
+      const newTokens = Object.values(newTokensMap);
+
+      const newDetectedTokens = detectedTokens.filter(
+        (token) => !importedTokensMap[token.address.toLowerCase()],
```

Comment:
This line is not covered. We don't have a test for adding a token that was previously detected.